theseus/labs/lie/functional/so3_impl.py

```diff
@@ -345,46 +347,50 @@ def _log_impl(group: torch.Tensor) -> torch.Tensor:
     # theta != pi
     near_zero_or_near_pi = torch.logical_or(near_zero, near_pi)
     # Compute the approximation of theta / sin(theta) when theta is near to 0
-    non_zero = torch.ones(1, dtype=group.dtype, device=group.device)
+    non_zero = group.new_ones(1)
     sine_nz = torch.where(near_zero_or_near_pi, non_zero, sine)
     scale = torch.where(
         near_zero_or_near_pi,
         1 + sine**2 / 6,
         theta / sine_nz,
     )
-    ret = sine_axis * scale.view(-1, 1)
+    ret = sine_axis * scale.view(batch + (1,))
 
     # # theta ~ pi
-    ddiag = torch.diagonal(group, dim1=1, dim2=2)
+    ddiag = torch.diagonal(group, dim1=-2, dim2=-1)
     # Find the index of major coloumns and diagonals
     major = torch.logical_and(
-        ddiag[:, 1] > ddiag[:, 0], ddiag[:, 1] > ddiag[:, 2]
-    ) + 2 * torch.logical_and(ddiag[:, 2] > ddiag[:, 0], ddiag[:, 2] > ddiag[:, 1])
-    aux = torch.ones(group.shape[0], dtype=torch.bool)
-    sel_rows = 0.5 * (group[aux, major] + group[aux, :, major])
-    sel_rows[aux, major] -= cosine
+        ddiag[..., 1] > ddiag[..., 0], ddiag[..., 1] > ddiag[..., 2]
+    ) + 2 * torch.logical_and(
+        ddiag[..., 2] > ddiag[..., 0], ddiag[..., 2] > ddiag[..., 1]
+    )
+    major = major.view(-1)
+    aux = torch.ones(batch, dtype=torch.bool)
+    sel_rows = 0.5 * (group[aux, major] + group[aux, :, major]).view(batch + (3,))
+    sel_rows[aux, major] -= cosine.view(-1)
     axis = sel_rows / torch.where(
         near_zero,
         non_zero,
-        sel_rows.norm(dim=1),
-    ).view(-1, 1)
-    sign_tmp = sine_axis[aux, major].sign()
+        sel_rows.norm(dim=-1),
+    ).view(batch + (1,))
+    sign_tmp = sine_axis[aux, major].sign().view(batch)
     sign = torch.where(sign_tmp != 0, sign_tmp, torch.ones_like(sign_tmp))
     tangent_vector = torch.where(
-        near_pi.view(-1, 1), axis * (theta * sign).view(-1, 1), ret
+        near_pi.view(batch + (1,)), axis * (theta * sign).view(batch + (1,)), ret
     )
 
     return tangent_vector
 
 
 def _jlog_impl(group: torch.Tensor) -> Tuple[List[torch.Tensor], torch.Tensor]:
     check_group_tensor(group)
-    sine_axis = group.new_zeros(group.shape[0], 3)
-    sine_axis[:, 0] = 0.5 * (group[:, 2, 1] - group[:, 1, 2])
-    sine_axis[:, 1] = 0.5 * (group[:, 0, 2] - group[:, 2, 0])
-    sine_axis[:, 2] = 0.5 * (group[:, 1, 0] - group[:, 0, 1])
-    cosine = 0.5 * (group[:, 0, 0] + group[:, 1, 1] + group[:, 2, 2] - 1)
-    sine = sine_axis.norm(dim=1)
+    batch = group.shape[:-2]
```

Comment:
As we discussed, it's better to remove all of this code duplication before making these changes. You had to make multiple changes in several places twice to get this working. And now you need to do it two more times for SE3. This is not scalable and error prone, harder to review, and harder to maintain. Let's fix this please.

Reply:
Added `_exp_impl_helper` and `_log_impl_helper` for simplification.